File: go/core/interfaces/rate_limiter.go
package interfaces

import "context"

// RateLimiter controls the rate of operations using a token bucket algorithm.
//
// Implementations: token bucket via golang.org/x/time/rate (default).
//
// All service code depends on this interface, never on a concrete rate
// limiter. Swap implementations via the ratelimiter.New factory.
type RateLimiter interface {
	// Allow reports whether an event may happen now (non-blocking).
	Allow() bool

	// Wait blocks until the limiter permits an event or ctx is cancelled.
	Wait(ctx context.Context) error
}

File: python/techai_webutils/src/techai_webutils/core/interfaces/rate_limiter.py
"""Rate limiter interface.

Mirrors Go's ``interfaces.RateLimiter``.
"""

from __future__ import annotations

from abc import ABC, abstractmethod


class RateLimiter(ABC):
    """Controls the rate of operations using a token bucket algorithm."""

    @abstractmethod
    def allow(self) -> bool:
        """Report whether an event may happen now (non-blocking)."""
        ...

    @abstractmethod
    async def wait(self) -> None:
        """Block until the limiter permits an event."""
        ...
